testr/packages.py

```diff
@@ -96,6 +95,35 @@ def fileno(self):
         return sys.stdout.fileno()
 
 
+def communicate(process, log=None, text=False):
+    """
+    Real-time reading of a subprocess stdout.
+    Parameters
+    ----------
+    process:
+        process returned by subprocess.Popen
+    log: a stream object
+    level: str or int
+        the logging level
+    text: bool
+        whether the process output is text or binary
+    """
+
+    while True:
+        if process.poll() is not None:
+            break
+        line = process.stdout.readline()
+        line = line if text else line.decode()
```

Comment:
I'm a little confused about the `text` option. For this application `text` is always false, so what sort of process might return a `str` instead of `bytes`? 

But more to the point, can't the type be auto-detected, either `isinstance(line, bytes)` or a try/except with decoding or looking for a `decode` attribute?

Reply:
Since python 3.7, `subprocess.Popen` has a `text` argument to tell it to open stdout as a string stream. Otherwise it is a binary stream, so it depends on how the process was called.

One option is to check `process.text_mode`.